tests/integration-tests/tests/cloudwatch_logging/test_cloudwatch_logging.py

```diff
@@ -0,0 +1,461 @@
+        status_cmd = "/opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a status"
+        status = json.loads(self._run_command_on_master(status_cmd))
+        self._cluster_log_state["MasterServer"]["agent_status"] = status["status"]
+
+    def _populate_compute_agent_status(self):
+        """Get the cloudwatch agent's status for all the compute nodes in the cluster."""
+        status_cmd = "/opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl {redirect} -a status"
+        compute_statuses = self._run_command_on_computes(status_cmd)
+        hostname_to_status_dict = {hostname: json.loads(status) for hostname, status in compute_statuses.items()}
+        for hostname, status_dict in hostname_to_status_dict.items():
+            self._cluster_log_state["ComputeFleet"][hostname]["agent_status"] = status_dict["status"]
+
+    def _populate_agent_status(self):
+        """Get the cloudwatch agent's status for all the nodes in the cluster."""
+        self._populate_master_agent_status()
+        self._populate_compute_agent_status()
+        LOGGER.debug("After populating agent statuses:\n{}".format(self._dump_cluster_log_state()))
+
+    def _set_cluster_log_state(self):
+        """
+        Get the state of the cluster as it pertains to the CloudWatch logging feature.
+
+        In particular:
+        * Identify which EC2 instances belong to this cluster
+        * Identify which logs are relevant to the MasterServer and ComputeFleet nodes
+        * Identify whether each of a node's relevant logs contain data or not. If they do contain
+          data, save the last line of the file.
+        * Get the CloudWatch agent's status for each node
+        """
+        self._get_initial_cluster_log_state()
+        self._get_relevant_logs()
+        self._populate_log_existence()
+        self._populate_log_emptiness_and_tail()
+        self._populate_agent_status()
+
+
+class Boto3DataRetriever:
+    """Class used to get data from CloudWatch logs."""
+
+    def __init__(self, region):
+        """Save references to the variables the class will need to get log data."""
+        self.region = region
+
+    def get_log_groups(self):
+        """Get list of log groups."""
+        logs_client = boto3.client('logs', region_name=self.region)
+        log_groups = logs_client.describe_log_groups()['logGroups']
+        LOGGER.debug("Log groups: {}\n".format(json.dumps(log_groups, indent=4)))
+        return log_groups
+
+    def get_log_streams(self, log_group_name):
+        """Get list of log streams."""
+        logs_client = boto3.client('logs', region_name=self.region)
+        streams = logs_client.describe_log_streams(logGroupName=log_group_name)['logStreams']
+        LOGGER.debug("Log streams for {group}:\n{streams}".format(
+            group=log_group_name, streams=json.dumps(streams, indent=4)))
+        return streams
+
+    def get_log_events(self, log_group_name, log_stream_name):
```

Comment:
What if there's 10,000,000 events?

Reply:
I didn't use an iterator for pagination because the log streams that I'm currently getting events from do not get big enough to exceed the 1MB limit enforced by the `get_log_events` method.

But now that I'm reconsidering the decision again, it probably makes sense to just do it now, given that it seems very possible that a log that does so might be added in the future.

Reply:
Actually now that I look at it again, the decision to ignore pagination was motivated by the fact that the data is only used to ensure:
* the log stream is not empty
* the last event in the corresponding log file on the instance has been ingested by the log stream

So if the assumption "less than ~8,000 log events will be ingested for any given log in-between the time the 'test tail event' is read and the time the log stream is queried for events" seems safe, I think this can probably remain as is.

Thoughts?

Reply:
Agreed